src/client/binary_client.cpp

```diff
@@ -455,30 +461,52 @@ namespace
       request.Header = CreateRequestHeader();
       request.Header.Timeout = 0; //We do not want the request to timeout!
 
-      ResponseCallback responseCallback = [this](std::vector<char> buffer){
+      ResponseCallback responseCallback = [this](std::vector<char> buffer, ResponseHeader h){
         if (Debug) {std::cout << "BinaryClient | Got Publish Response, from server " << std::endl;}
-        BufferInputChannel bufferInput(buffer);
-        IStreamBinary in(bufferInput);
-        PublishResponse response;
-        in >> response;
+		PublishResponse response;
+		if (h.ServiceResult != OpcUa::StatusCode::Good)
+		{
+			response.Header = std::move(h);
+		}
+		else
+		{
+			BufferInputChannel bufferInput(buffer);
+			IStreamBinary in(bufferInput);
+			in >> response;
+		}
+        
         CallbackService.post([this, response]() 
             { 
-              if (Debug) {std::cout << "BinaryClient | Calling callback for Subscription " << response.Result.SubscriptionId << std::endl;}
-              SubscriptionCallbackMap::const_iterator callbackIt = this->PublishCallbacks.find(response.Result.SubscriptionId);
-              if (callbackIt == this->PublishCallbacks.end())
-              {
-                std::cout << "BinaryClient | Error Uknown SubscriptionId " << response.Result.SubscriptionId << std::endl;
-              }
-              else
-              {
-                try { //calling client code, better put it under try/catch otherwise we crash entire client
-                  callbackIt->second(response.Result);
-                }
-                catch (const std::exception& ex)
-                {
-                  std::cout << "Error calling application callback " << ex.what() << std::endl;
-                }
-              }
+			if (response.Header.ServiceResult == OpcUa::StatusCode::Good)
+			{
+				if (Debug) { std::cout << "BinaryClient | Calling callback for Subscription " << response.Result.SubscriptionId << std::endl; }
+				SubscriptionCallbackMap::const_iterator callbackIt = this->PublishCallbacks.find(response.Result.SubscriptionId);
+				if (callbackIt == this->PublishCallbacks.end())
+				{
+					std::cout << "BinaryClient | Error Unknown SubscriptionId " << response.Result.SubscriptionId << std::endl;
+				}
+				else
+				{
+					try { //calling client code, better put it under try/catch otherwise we crash entire client
+						callbackIt->second(response.Result);
+					}
+					catch (const std::exception& ex)
+					{
+						std::cout << "Error calling application callback " << ex.what() << std::endl;
+					}
+				}
+			}
+			else if (response.Header.ServiceResult == OpcUa::StatusCode::BadSessionClosed)
+			{
+				if (Debug) 
+				{
+					std::cout << "BinaryClient | Session is closed";
+				}
+			}
+			else
+			{
+				// TODO
```

Comment:
Can You print a big warning here? We must let users know the server has an issue. Even better we should raise an exception in client code-